File: python/src/wgrok/listener.py
"""Platform listener abstraction — normalized message receiving across transports.

Each platform listener connects via its native mechanism (WebSocket, TCP, etc.)
and emits messages through a common callback interface.
"""

from __future__ import annotations

import asyncio
import contextlib
import json
import re
from dataclasses import dataclass
from typing import Any

import aiohttp
from webex_message_handler import WebexMessageHandler, WebexMessageHandlerConfig

from .logging import NdjsonLogger, NoopLogger

Logger = NdjsonLogger | NoopLogger


@dataclass
class IncomingMessage:
    """Normalized incoming message from any platform."""

    sender: str
    text: str
    msg_id: str
    platform: str
    cards: list[dict]


MessageCallback = Any  # Callable[[IncomingMessage], Awaitable[None]]


class PlatformListener:
    """Base class for platform listeners."""

    async def connect(self) -> None:
        raise NotImplementedError

    async def disconnect(self) -> None:
        raise NotImplementedError

    def on_message(self, callback: MessageCallback) -> None:
        raise NotImplementedError


class WebexListener(PlatformListener):
    """Webex listener using webex-message-handler (Mercury WebSocket + KMS)."""

    def __init__(self, token: str, logger: Logger) -> None:
        self._token = token
        self._logger = logger
        self._handler: WebexMessageHandler | None = None
        self._callback: MessageCallback | None = None

    def on_message(self, callback: MessageCallback) -> None:
        self._callback = callback

    async def connect(self) -> None:
        config = WebexMessageHandlerConfig(token=self._token, logger=self._logger)
        self._handler = WebexMessageHandler(config)

        @self._handler.on("message:created")
        async def _on_msg(msg) -> None:
            if self._callback:
                incoming = IncomingMessage(
                    sender=msg.person_email,
                    text=(msg.text or "").strip(),
                    msg_id=msg.id,
                    platform="webex",
                    cards=[],
                )
                await self._callback(incoming)

        await self._handler.connect()

    async def disconnect(self) -> None:
        if self._handler:
            await self._handler.disconnect()
            self._handler = None


SLACK_SOCKET_MODE_URL = "https://slack.com/api/apps.connections.open"


class SlackListener(PlatformListener):
    """Slack listener using Socket Mode WebSocket.

    Requires an app-level token (xapp-*) for the WebSocket connection.
    """

    def __init__(self, token: str, logger: Logger) -> None:
        self._token = token
        self._logger = logger
        self._callback: MessageCallback | None = None
        self._ws: aiohttp.ClientWebSocketResponse | None = None
        self._session: aiohttp.ClientSession | None = None
        self._running = False

    def on_message(self, callback: MessageCallback) -> None:
        self._callback = callback

    async def connect(self) -> None:
        self._session = aiohttp.ClientSession()

        # Request a WebSocket URL via apps.connections.open
        async with self._session.post(
            SLACK_SOCKET_MODE_URL,
            headers={"Authorization": f"Bearer {self._token}"},
        ) as resp:
            resp.raise_for_status()
            data = await resp.json()
            if not data.get("ok"):
                raise ConnectionError(f"Slack apps.connections.open failed: {data.get('error', 'unknown')}")
            ws_url = data["url"]

        self._ws = await self._session.ws_connect(ws_url)
        self._running = True
        self._logger.info("Slack Socket Mode connected")

        # Start reading messages in background
        asyncio.create_task(self._read_loop())

    async def _read_loop(self) -> None:
        """Read Socket Mode events and dispatch message events."""
        while self._running and self._ws and not self._ws.closed:
            try:
                msg = await self._ws.receive()
            except Exception:
                break

            if msg.type == aiohttp.WSMsgType.TEXT:
                await self._handle_event(msg.data)
            elif msg.type in (aiohttp.WSMsgType.CLOSED, aiohttp.WSMsgType.ERROR):
                break

    async def _handle_event(self, raw: str) -> None:
        """Handle a Socket Mode envelope."""
        try:
            envelope = json.loads(raw)
        except json.JSONDecodeError:
            return

        # Acknowledge the envelope
        envelope_id = envelope.get("envelope_id")
        if envelope_id and self._ws and not self._ws.closed:
            await self._ws.send_json({"envelope_id": envelope_id})

        event_type = envelope.get("type")
        if event_type != "events_api":
            return

        payload = envelope.get("payload", {})
        event = payload.get("event", {})
        if event.get("type") != "message":
            return

        # Skip bot messages to avoid loops
        if event.get("bot_id"):
            return

        if self._callback:
            incoming = IncomingMessage(
                sender=event.get("user", ""),
                text=(event.get("text", "") or "").strip(),
                msg_id=event.get("ts", ""),
                platform="slack",
                cards=[],
            )
            await self._callback(incoming)

    async def disconnect(self) -> None:
        self._running = False
        if self._ws and not self._ws.closed:
            await self._ws.close()
            self._ws = None
        if self._session:
            await self._session.close()
            self._session = None
        self._logger.info("Slack listener disconnected")


DISCORD_GATEWAY_URL = "wss://gateway.discord.gg/?v=10&encoding=json"
DISCORD_GATEWAY_API = "https://discord.com/api/v10/gateway"

# Discord Gateway opcodes
_OP_DISPATCH = 0
_OP_HEARTBEAT = 1
_OP_IDENTIFY = 2
_OP_HELLO = 10
_OP_HEARTBEAT_ACK = 11

# Intents: GUILD_MESSAGES (1 << 9) + MESSAGE_CONTENT (1 << 15)
_INTENTS = (1 << 9) | (1 << 15)


class DiscordListener(PlatformListener):
    """Discord listener using Gateway WebSocket."""

    def __init__(self, token: str, logger: Logger) -> None:
        self._token = token
        self._logger = logger
        self._callback: MessageCallback | None = None
        self._ws: aiohttp.ClientWebSocketResponse | None = None
        self._session: aiohttp.ClientSession | None = None
        self._running = False
        self._heartbeat_task: asyncio.Task | None = None
        self._sequence: int | None = None

    def on_message(self, callback: MessageCallback) -> None:
        self._callback = callback

    async def connect(self) -> None:
        self._session = aiohttp.ClientSession()

        # Get gateway URL
        async with self._session.get(DISCORD_GATEWAY_API) as resp:
            resp.raise_for_status()
            data = await resp.json()
            gw_url = data.get("url", "wss://gateway.discord.gg")

        ws_url = f"{gw_url}/?v=10&encoding=json"
        self._ws = await self._session.ws_connect(ws_url)
        self._running = True

        # Wait for Hello (opcode 10)
        hello_msg = await self._ws.receive_json()
        if hello_msg.get("op") != _OP_HELLO:
            raise ConnectionError(f"Expected Hello (op 10), got op {hello_msg.get('op')}")

        heartbeat_interval = hello_msg["d"]["heartbeat_interval"] / 1000.0

        # Start heartbeat
        self._heartbeat_task = asyncio.create_task(self._heartbeat_loop(heartbeat_interval))

        # Send Identify
        await self._ws.send_json({
            "op": _OP_IDENTIFY,
            "d": {
                "token": self._token,
                "intents": _INTENTS,
                "properties": {
                    "os": "linux",
                    "browser": "wgrok",
                    "device": "wgrok",
                },
            },
        })

        self._logger.info("Discord Gateway connected")

        # Start reading events in background
        asyncio.create_task(self._read_loop())

    async def _heartbeat_loop(self, interval: float) -> None:
        """Send periodic heartbeats to keep the connection alive."""
        while self._running:
            await asyncio.sleep(interval)
            if self._ws and not self._ws.closed:
                await self._ws.send_json({"op": _OP_HEARTBEAT, "d": self._sequence})

    async def _read_loop(self) -> None:
        """Read Gateway events and dispatch MESSAGE_CREATE."""
        while self._running and self._ws and not self._ws.closed:
            try:
                msg = await self._ws.receive()
            except Exception:
                break

            if msg.type == aiohttp.WSMsgType.TEXT:
                try:
                    data = json.loads(msg.data)
                except json.JSONDecodeError:
                    continue

                # Track sequence number for heartbeat
                if data.get("s") is not None:
                    self._sequence = data["s"]

                op = data.get("op")
                if op == _OP_DISPATCH and data.get("t") == "MESSAGE_CREATE":
                    await self._handle_message_create(data.get("d", {}))

            elif msg.type in (aiohttp.WSMsgType.CLOSED, aiohttp.WSMsgType.ERROR):
                break

    async def _handle_message_create(self, event: dict) -> None:
        """Handle a MESSAGE_CREATE dispatch event."""
        # Skip bot messages
        author = event.get("author", {})
        if author.get("bot"):
            return

        if self._callback:
            # Build embeds list as "cards"
            embeds = event.get("embeds", [])
            incoming = IncomingMessage(
                sender=author.get("id", ""),
                text=(event.get("content", "") or "").strip(),
                msg_id=event.get("id", ""),
                platform="discord",
                cards=embeds,
            )
            await self._callback(incoming)

    async def disconnect(self) -> None:
        self._running = False
        if self._heartbeat_task:
            self._heartbeat_task.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await self._heartbeat_task
            self._heartbeat_task = None
        if self._ws and not self._ws.closed:
            await self._ws.close()
            self._ws = None
        if self._session:
            await self._session.close()
            self._session = None
        self._logger.info("Discord listener disconnected")


_PRIVMSG_RE = re.compile(r"^:([^!]+)![^ ]+ PRIVMSG ([^ ]+) :(.+)$")


class IrcListener(PlatformListener):
    """IRC listener using persistent TCP/TLS connection.

    Parses incoming PRIVMSG lines and emits them as normalized messages.
    """

    def __init__(self, conn_str: str, logger: Logger) -> None:
        from .irc import IrcConnection

        self._conn = IrcConnection(conn_str)
        self._logger = logger
        self._callback: MessageCallback | None = None
        self._running = False

    def on_message(self, callback: MessageCallback) -> None:
        self._callback = callback

    async def connect(self) -> None:
        await self._conn.connect()
        self._running = True
        self._logger.info(f"IRC connected to {self._conn.nick}@{self._conn.channel}")
        asyncio.create_task(self._read_loop())

    async def _read_loop(self) -> None:
        """Read IRC lines and dispatch PRIVMSG events."""
        reader = self._conn._reader
        while self._running and reader:
            try:
                line = await asyncio.wait_for(reader.readline(), timeout=300)
            except asyncio.TimeoutError:
                # Send a PING to keep connection alive
                if self._conn.connected:
                    await self._conn._send_raw("PING :keepalive")
                continue
            except Exception:
                break

            if not line:
                break

            decoded = line.decode("utf-8", errors="replace").strip()
            if not decoded:
                continue

            # Handle server PING
            if decoded.startswith("PING"):
                pong_arg = decoded[5:] if len(decoded) > 5 else ""
                await self._conn._send_raw(f"PONG {pong_arg}")
                continue

            # Parse PRIVMSG
            match = _PRIVMSG_RE.match(decoded)
            if match and self._callback:
                nick, target, text = match.groups()
                incoming = IncomingMessage(
                    sender=nick,
                    text=text.strip(),
                    msg_id="",
                    platform="irc",
                    cards=[],
                )
                await self._callback(incoming)

    async def disconnect(self) -> None:
        self._running = False
        await self._conn.disconnect()
        self._logger.info("IRC listener disconnected")


def create_listener(platform: str, token: str, logger: Logger) -> PlatformListener:
    """Factory: create the right listener for a platform."""
    if platform == "webex":
        return WebexListener(token, logger)
    if platform == "slack":
        return SlackListener(token, logger)
    if platform == "discord":
        return DiscordListener(token, logger)
    if platform == "irc":
        return IrcListener(token, logger)
    raise ValueError(f"Unsupported platform: {platform}")

File: ts/src/listener.ts
/**
 * Platform listener abstraction — normalized message receiving across transports.
 *
 * Each platform listener connects via its native mechanism (WebSocket, TCP, etc.)
 * and emits messages through a common callback interface.
 */

import { WebexMessageHandler, type Logger } from 'webex-message-handler';
import type { TLSSocket } from 'tls';
import * as tls from 'tls';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type DecryptedMessage = any;
type SlackEnvelope = Record<string, unknown>;
type SlackPayload = Record<string, unknown>;
type SlackEvent = Record<string, unknown>;
type DiscordMessage = Record<string, unknown>;
type DiscordData = Record<string, unknown>;

export interface IncomingMessage {
  sender: string;
  text: string;
  html: string;
  msgId: string;
  platform: string;
  cards: unknown[];
}

export type MessageCallback = (msg: IncomingMessage) => Promise<void> | void;

export interface PlatformListener {
  onMessage(callback: MessageCallback): void;
  connect(): Promise<void>;
  disconnect(): Promise<void>;
}

/**
 * WebexListener wraps webex-message-handler and normalizes into IncomingMessage
 */
export class WebexListener implements PlatformListener {
  private token: string;
  private logger: Logger;
  private callback: MessageCallback | null = null;
  private handler: WebexMessageHandler | null = null;

  constructor(token: string, logger: Logger) {
    this.token = token;
    this.logger = logger;
  }

  onMessage(callback: MessageCallback): void {
    this.callback = callback;
  }

  async connect(): Promise<void> {
    this.handler = new WebexMessageHandler({
      token: this.token,
      logger: this.logger,
    });

    this.handler.on('message:created', (msg) => {
      void this.handleWebexMessage(msg);
    });

    await this.handler.connect();
  }

  private async handleWebexMessage(msg: DecryptedMessage): Promise<void> {
    if (this.callback) {
      const incoming: IncomingMessage = {
        sender: msg.personEmail as string,
        text: ((msg.text as string) || '').trim(),
        html: ((msg.html as string) || ''),
        msgId: msg.id as string,
        platform: 'webex',
        cards: [],
      };
      await this.callback(incoming);
    }
  }

  async disconnect(): Promise<void> {
    if (this.handler) {
      await this.handler.disconnect();
      this.handler = null;
    }
  }
}

const SLACK_SOCKET_MODE_URL = 'https://slack.com/api/apps.connections.open';

/**
 * SlackListener uses Socket Mode WebSocket
 * Requires an app-level token (xapp-*)
 */
export class SlackListener implements PlatformListener {
  private token: string;
  private logger: Logger;
  private callback: MessageCallback | null = null;
  private ws: WebSocket | null = null;
  private running = false;

  constructor(token: string, logger: Logger) {
    this.token = token;
    this.logger = logger;
  }

  onMessage(callback: MessageCallback): void {
    this.callback = callback;
  }

  async connect(): Promise<void> {
    // Request a WebSocket URL via apps.connections.open
    const response = await fetch(SLACK_SOCKET_MODE_URL, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${this.token}`,
      },
    });

    if (!response.ok) {
      throw new Error(`Slack apps.connections.open failed: HTTP ${response.status}`);
    }

    const data = (await response.json()) as Record<string, unknown>;
    if (!data.ok) {
      throw new Error(`Slack apps.connections.open failed: ${(data.error as string | undefined) || 'unknown'}`);
    }

    const wsUrl = data.url as string;
    this.ws = new WebSocket(wsUrl);
    this.running = true;

    this.ws.addEventListener('message', (event) => {
      void this.handleSlackEvent(event.data as string);
    });

    this.ws.addEventListener('close', () => {
      this.running = false;
    });

    this.ws.addEventListener('error', () => {
      this.running = false;
    });

    this.logger.info('Slack Socket Mode connected');

    // Wait for the WebSocket to be open before returning
    await new Promise<void>((resolve) => {
      if (this.ws?.readyState === WebSocket.OPEN) {
        resolve();
      } else {
        const onOpen = () => {
          this.ws?.removeEventListener('open', onOpen);
          resolve();
        };
        this.ws?.addEventListener('open', onOpen);
      }
    });
  }

  private async handleSlackEvent(raw: string): Promise<void> {
    let envelope: SlackEnvelope;
    try {
      envelope = JSON.parse(raw) as SlackEnvelope;
    } catch {
      return;
    }

    // Acknowledge the envelope
    const envelopeId = envelope.envelope_id as string | undefined;
    if (envelopeId && this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify({ envelope_id: envelopeId }));
    }

    const eventType = envelope.type as string | undefined;
    if (eventType !== 'events_api') {
      return;
    }

    const payload = envelope.payload as SlackPayload | undefined;
    if (!payload) {
      return;
    }

    const event = payload.event as SlackEvent | undefined;
    if (!event || (event.type as string) !== 'message') {
      return;
    }

    // Skip bot messages to avoid loops
    if (event.bot_id) {
      return;
    }

    if (this.callback) {
      const incoming: IncomingMessage = {
        sender: (event.user as string) || '',
        text: (((event.text as string) || '').trim()),
        html: '',
        msgId: (event.ts as string) || '',
        platform: 'slack',
        cards: [],
      };
      await this.callback(incoming);
    }
  }

  async disconnect(): Promise<void> {
    this.running = false;
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
    this.logger.info('Slack listener disconnected');
  }
}

const DISCORD_GATEWAY_API = 'https://discord.com/api/v10/gateway';

// Discord Gateway opcodes
const OP_DISPATCH = 0;
const OP_HEARTBEAT = 1;
const OP_IDENTIFY = 2;
const OP_HELLO = 10;

// Intents: GUILD_MESSAGES (1 << 9) + MESSAGE_CONTENT (1 << 15)
const INTENTS = (1 << 9) | (1 << 15);

/**
 * DiscordListener uses Gateway WebSocket
 */
export class DiscordListener implements PlatformListener {
  private token: string;
  private logger: Logger;
  private callback: MessageCallback | null = null;
  private ws: WebSocket | null = null;
  private running = false;
  private heartbeatInterval: NodeJS.Timeout | null = null;
  private sequence: number | null = null;

  constructor(token: string, logger: Logger) {
    this.token = token;
    this.logger = logger;
  }

  onMessage(callback: MessageCallback): void {
    this.callback = callback;
  }

  async connect(): Promise<void> {
    // Get gateway URL
    const response = await fetch(DISCORD_GATEWAY_API);
    if (!response.ok) {
      throw new Error(`Failed to get Discord gateway: HTTP ${response.status}`);
    }

    const data = (await response.json()) as Record<string, unknown>;
    const gwUrl = (data.url as string) || 'wss://gateway.discord.gg';
    const wsUrl = `${gwUrl}/?v=10&encoding=json`;

    this.ws = new WebSocket(wsUrl);
    this.running = true;

    this.ws.addEventListener('close', () => {
      this.running = false;
      if (this.heartbeatInterval) {
        clearInterval(this.heartbeatInterval);
        this.heartbeatInterval = null;
      }
    });

    this.ws.addEventListener('error', () => {
      this.running = false;
    });

    // Wait for Hello (opcode 10), then register the main message handler
    await new Promise<void>((resolve, reject) => {
      const onMessage = async (event: Event) => {
        if (event instanceof MessageEvent) {
          try {
            const helloMsg = JSON.parse(event.data as string) as Record<string, unknown>;
            if ((helloMsg.op as number) === OP_HELLO) {
              this.ws?.removeEventListener('message', onMessage);
              this.ws?.removeEventListener('error', onError);
              const heartbeatInterval = ((helloMsg.d as Record<string, unknown>).heartbeat_interval as number) / 1000;
              this.startHeartbeat(heartbeatInterval);
              resolve();
            }
          } catch {
            reject(new Error('Failed to parse Discord Hello message'));
          }
        }
      };

      const onError = () => {
        this.ws?.removeEventListener('message', onMessage);
        reject(new Error('WebSocket error during Hello'));
      };

      this.ws?.addEventListener('message', onMessage);
      this.ws?.addEventListener('error', onError);
    });

    // Register the main dispatch handler after Hello is complete
    this.ws?.addEventListener('message', (event) => {
      void this.handleDiscordMessage(event.data as string);
    });

    // Send Identify
    this.ws?.send(
      JSON.stringify({
        op: OP_IDENTIFY,
        d: {
          token: this.token,
          intents: INTENTS,
          properties: {
            os: 'linux',
            browser: 'wgrok',
            device: 'wgrok',
          },
        },
      })
    );

    this.logger.info('Discord Gateway connected');
  }

  private startHeartbeat(interval: number): void {
    this.heartbeatInterval = setInterval(() => {
      if (this.ws && this.ws.readyState === WebSocket.OPEN) {
        this.ws.send(JSON.stringify({ op: OP_HEARTBEAT, d: this.sequence }));
      }
    }, interval * 1000);
  }

  private async handleDiscordMessage(raw: string): Promise<void> {
    let data: DiscordMessage;
    try {
      data = JSON.parse(raw) as DiscordMessage;
    } catch {
      return;
    }

    // Track sequence number for heartbeat
    if ((data.s as number | null) !== null && data.s !== undefined) {
      this.sequence = data.s as number;
    }

    const op = data.op as number | undefined;
    if (op === OP_DISPATCH && (data.t as string) === 'MESSAGE_CREATE') {
      await this.handleDiscordMessageCreate(data.d as DiscordData);
    }
  }

  private async handleDiscordMessageCreate(event: DiscordData): Promise<void> {
    // Skip bot messages
    const author = event.author as Record<string, unknown> | undefined;
    if (!author || (author.bot as boolean | undefined)) {
      return;
    }

    if (this.callback) {
      const embeds = event.embeds as unknown[] | undefined;
      const incoming: IncomingMessage = {
        sender: (author.id as string) || '',
        text: (((event.content as string) || '').trim()),
        html: '',
        msgId: (event.id as string) || '',
        platform: 'discord',
        cards: embeds || [],
      };
      await this.callback(incoming);
    }
  }

  async disconnect(): Promise<void> {
    this.running = false;
    if (this.heartbeatInterval) {
      clearInterval(this.heartbeatInterval);
      this.heartbeatInterval = null;
    }
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
    this.logger.info('Discord listener disconnected');
  }
}

const PRIVMSG_RE = /^:([^!]+)![^ ]+ PRIVMSG ([^ ]+) :(.+)$/;

/**
 * IrcListener uses persistent TCP/TLS connection
 */
export class IrcListener implements PlatformListener {
  private connStr: string;
  private logger: Logger;
  private callback: MessageCallback | null = null;
  private socket: TLSSocket | null = null;
  private running = false;
  private buffer = '';

  constructor(connStr: string, logger: Logger) {
    this.connStr = connStr;
    this.logger = logger;
  }

  onMessage(callback: MessageCallback): void {
    this.callback = callback;
  }

  async connect(): Promise<void> {
    const params = this.parseConnectionString(this.connStr);

    // Create TLS connection
    this.socket = tls.connect(
      {
        host: params.server,
        port: params.port,
        rejectUnauthorized: false,
      },
      () => {
        void this.sendRaw(`NICK ${params.nick}`);
        if (params.password) {
          void this.sendRaw(`PASS ${params.password}`);
        }
        void this.sendRaw(`USER ${params.nick} 0 * :${params.nick}`);
        void this.waitForWelcome(params.nick, params.channel);
      }
    );

    this.socket.on('data', (data: Buffer | string) => {
      this.handleIrcData(data);
    });

    this.socket.on('close', () => {
      this.running = false;
    });

    this.socket.on('error', () => {
      this.running = false;
    });

    this.running = true;
    this.logger.info(`IRC connected to ${params.nick}@${params.server}`);
  }

  private parseConnectionString(connStr: string): {
    nick: string;
    password: string;
    server: string;
    port: number;
    channel: string;
  } {
    if (!connStr.includes('@')) {
      throw new Error(`Invalid IRC connection string (missing @): ${connStr}`);
    }

    const [creds, serverPart] = connStr.split('@', 2);

    let nick: string;
    let password: string;
    if (creds.includes(':')) {
      [nick, password] = creds.split(':', 2);
    } else {
      nick = creds;
      password = '';
    }

    let hostPort: string;
    let channel: string;
    if (serverPart.includes('/')) {
      [hostPort, channel] = serverPart.split('/', 2);
    } else {
      hostPort = serverPart;
      channel = '';
    }

    let server: string;
    let port: number;
    if (hostPort.includes(':')) {
      const parts = hostPort.split(':');
      server = parts.slice(0, -1).join(':');
      const portStr = parts[parts.length - 1];
      port = parseInt(portStr, 10);
    } else {
      server = hostPort;
      port = 6697; // Default TLS port
    }

    return { nick, password, server, port, channel };
  }

  private async sendRaw(line: string): Promise<void> {
    if (!this.socket) {
      throw new Error('Not connected to IRC server');
    }
    // Sanitize to prevent IRC protocol injection
    const sanitized = line.replace(/\r/g, '').replace(/\n/g, '');
    return new Promise((resolve, reject) => {
      this.socket!.write(`${sanitized}\r\n`, (err) => {
        if (err) reject(err);
        else resolve();
      });
    });
  }

  private async waitForWelcome(nick: string, channel: string): Promise<void> {
    // Welcome is async, just wait a bit then join the channel
    await new Promise((resolve) => setTimeout(resolve, 1000));
    if (channel && this.socket) {
      void this.sendRaw(`JOIN ${channel}`);
    }
  }

  private handleIrcData(data: Buffer | string): void {
    this.buffer += typeof data === 'string' ? data : data.toString('utf-8');

    let lines = this.buffer.split('\r\n');
    this.buffer = lines[lines.length - 1];
    lines = lines.slice(0, -1);

    for (const line of lines) {
      const decoded = line.trim();
      if (!decoded) {
        continue;
      }

      // Handle server PING
      if (decoded.startsWith('PING')) {
        const pongArg = decoded.substring(5).trim();
        void this.sendRaw(`PONG ${pongArg}`);
        continue;
      }

      // Parse PRIVMSG
      const match = PRIVMSG_RE.exec(decoded);
      if (match && this.callback) {
        const [, nick, , text] = match;
        void (async () => {
          const incoming: IncomingMessage = {
            sender: nick,
            text: text.trim(),
            html: '',
            msgId: '',
            platform: 'irc',
            cards: [],
          };
          await this.callback!(incoming);
        })();
      }
    }
  }

  async disconnect(): Promise<void> {
    this.running = false;
    if (this.socket) {
      this.socket.end();
      this.socket = null;
    }
    this.logger.info('IRC listener disconnected');
  }
}

export function createListener(platform: string, token: string, logger: Logger): PlatformListener {
  if (platform === 'webex') {
    return new WebexListener(token, logger);
  }
  if (platform === 'slack') {
    return new SlackListener(token, logger);
  }
  if (platform === 'discord') {
    return new DiscordListener(token, logger);
  }
  if (platform === 'irc') {
    return new IrcListener(token, logger);
  }
  throw new Error(`Unsupported platform: ${platform}`);
}
